### PR TITLE
Port the Neovim popup menu support for command-line completion

### PR DESCRIPTION
**Pull request by yegappan**
The following Neovim patches are part of this PR:

Author: Björn Linse <bjorn.linse@gmail.com>
Commit: https://github.com/neovim/neovim/commit/be8ebba325451b387c0aedacfcda6c53e6c51188

Author: Björn Linse <bjorn.linse@gmail.com>
Commit: https://github.com/neovim/neovim/commit/0636b25f28e408c8b16026354db7edfef079614a

Author: Billy Su <g4691821@gmail.com>
Commit: https://github.com/neovim/neovim/commit/961c528afc7f3a21a8c4b118a2c585dbe670c5c5

Author: akovaski <akaskik@gmail.com>
Commit: https://github.com/neovim/neovim/commit/97dcc48c998ccecaa37a3cbea568d85c2f1407f9

Author: glacambre <code@lacamb.re>
Commit: https://github.com/neovim/neovim/commit/958ca938b44ea7d6f126887f5bec98ee6d2f6f15

The following Neovim patch is not applied as the Vim screen redraw code is different and the popup
menu is already correctly positioned:

Author: Björn Linse <bjorn.linse@gmail.com>
Commit: https://github.com/neovim/neovim/commit/b98d1cf68bd1d78d8170a8d994ba7dae5540962a

If we run into any problems related to positioning the popup menu, then we need to revisit this patch.

### CODE REVIEW

**Comment by brammool** [comment]
Thanks for working on this.  The number of changes is smaller than what I expected.

The navigation with CTRL-N and CTRL-P also works, which is the most natural for me.  This should be mentioned somewhere.

The navigation with cursor keys is weird.  Up and Down should really select the next/previous entry.  Left and Right can be used for what they used to do, thus swap them.

It appears the menu is not positioned properly. E.g. ":e proto/<Tab>" shows the menu too far left. With ":e proto/p<Tab> it's off by one character.  Can we align it with the text in the command line?  It won't always fit, but should work in most cases.

When there are many matches there is a scrollbar, trying to use the mouse scrolls the whole window.  Perhaps actual scrolling the menu isn't implemented, but it should probably disable the mouse scrolling.

Would be good to have a couple of screenshot tests, especially for the positioning.

File: src/cmdexpand.c [hunk 1]
```diff
@@ -553,6 +604,28 @@ showmatches(expand_T *xp, int wildmenu UNUSED)
...
+	compl_match_arraysize = num_files;
+	compl_match_array = ALLOC_MULT(pumitem_T, compl_match_arraysize);
+	for (i = 0; i < num_files; i++)
+	{
+	    compl_match_array[i].pum_text = L_SHOWFILE(i);
+	    compl_match_array[i].pum_info = NULL;
+	    compl_match_array[i].pum_extra = NULL;
+	    compl_match_array[i].pum_kind = NULL;
+	}
+	endpos = (showtail ? sm_gettail(xp->xp_pattern) : xp->xp_pattern);
```
Comment (dpelle) [inline src/cmdexpand.c]: (nitpicky) the parentheses are not useful here.

**Comment by bfrg** [comment]
It doesn't work with `set wildoptions=pum,tagfile` or `set wildoptions=tagfile,pum`.

**Comment by yegappan** [comment]
>
> It doesn't work with `set wildoptions=pum,tagfile` or `set wildoptions=tagfile,pum`.
>

This is addressed by the latest commit.

File: src/cmdexpand.c (hunk 1 above)
Comment (yegappan) [reply src/cmdexpand.c]: Thanks. This is addressed.

**Comment by yegappan** [comment]
> 
> The navigation with CTRL-N and CTRL-P also works, which is the most natural for me. This should be mentioned somewhere.
>

Updated the help text to mention these keys.

> 
> The navigation with cursor keys is weird. Up and Down should really select the next/previous entry. Left and Right can
> be used for what they used to do, thus swap them.

I have swapped these keys now. Hopefully this doesn't cause new issues.

> 
> It appears the menu is not positioned properly. E.g. ":e proto/" shows the menu too far left. With ":e proto/p it's off by
> one character. Can we align it with the text in the command line? It won't always fit, but should work in most cases.
>

Changed the popup menu position to the end of the cmd line. I am not sure whether this works in all the cases.

> 
> When there are many matches there is a scrollbar, trying to use the mouse scrolls the whole window. Perhaps actual
> scrolling the menu isn't implemented, but it should probably disable the mouse scrolling.
>

Need to look into this.

>
> Would be good to have a couple of screenshot tests, especially for the positioning.
>

Yes. I will add some screen shot tests.

**Comment by bfrg** [comment]
When I press `:e <tab>` in command-line mode, the popup menu opens as expected, but after pressing `<c-e>` to cancel the pum, followed by `<up>`, Vim will insert `../` into the command-line, open the popup menu and select the first entry. This is unexpected. `<up>` and `<down>` keys should just navigate through the command-line history and not open the popup menu.

Is the pum aligned correctly in this screenshot? It seems to be off by 2 characters.
screenshot-2022-02-08_050536

**Comment by vim-ml** [comment]
Hi,

On Mon, Feb 7, 2022 at 8:13 PM bfrg ***@***.***> wrote:

> When I press :e <tab> in command-line mode, the popup menu opens as
> expected, but after pressing <c-e> to cancel the pum, followed by <up>,
> Vim will insert ../ into the command-line, open the popup menu and select
> the first entry. This is unexpected. <up> and <down> keys should just
> navigate through the command-line history and not open the popup menu.
>

Thanks for trying out the patch and reporting this issue. I have updated
the PR to fix this issue.
Can you try out the updated PR?


> Is the pum aligned correctly in this screenshot? It seems to be off by 2
> characters.
> [image: screenshot-2022-02-08_050536]
> <https://user-images.githubusercontent.com/6266600/152916289-04353212-5d46-4684-97f7-f4d3abc460ab.png>
>
>
> I think you typed "cN" and then pressed <Tab> to display the popup menu.
In that case, the popup menu
is correctly aligned. The popup menu uses the last column in the cmd line
where you invoked the completion.

Regards,
Yegappan

**Comment by puremourning** [comment]
What does `pumvisible()` return when this popup is displayed? I'm thinking that a lot of uses have mappings for `<Tab>` that look like `inoremap <expr> <Tab> pumvisible() ? '<C-n>' : '<Tab>'`. I can imagine a use case for that as a command line mapping too (i.e. to make Tab work like Ctrl-n when the menu is visible). WDYT?

**Comment by brammool** [comment]
Let me include this now, so users can try it out.  I expect there will be some improvements later.

**Comment by brammool** [comment]
I'll fix the horizontal positioning.

**Comment by bfrg** [comment]
@puremourning `<tab>` already works out of the box, `pumvisible()` too. The only difference I found is that `<up>` and `<down>` keys select the previous and next entry, respectively, and insert it similar to `<c-p>` and `<c-n>`. This is different from insert-mode completion where `<up>` and `<down>` will navigate through the menu without inserting the entry.

**Comment by puremourning** [comment]
> <up> and <down> will navigate through the menu without inserting the entry.

Only if `completeopt` contains `noinsert`  right?

**Comment by habamax** [comment]
Huh, I take my words back, it works for everything, indeed. For buffers I tried the old vim.

**Comment by Shane-XB-Qian** [comment]
> Only if `completeopt` contains `noinsert` right?

those `opt` options looks not worked here. 
it's more like `wildmenu`, but just showing at 'popup menu' vs 'stl'.

**Comment by bfrg** [comment]
@puremourning `<up>` and `<down>` never insert anything, they just select the previous or next entries in the menu. This is explained under `:h popupmenu-keys`.

Like @Shane-XB-Qian said, the new `pum` option behaves exactly like `wildmenu` but using a popup menu instead of the statusline.

`pumvisible()` can be used. For example, the following works:
```vim
cnoremap <expr> <enter> pumvisible() ? '<c-y>' : '<enter>'
cnoremap <expr> <esc>   pumvisible() ? '<c-e>' : '<esc>'
```

**Comment by bfrg** [comment]
Hmm, It seems the second mappings doesn't work after all:
```vim
cnoremap <expr> <esc> pumvisible() ? '<c-e>' : '<esc>'
```

**Comment by Shane-XB-Qian** [comment]
no, it worked.
// but wildmode=list,full not supported, and wildignorecase not worked for all case ignore, and wildmenumode() seems might be wrong.

**Comment by Shane-XB-Qian** [comment]
no, looks it worked.
// but wildmode=list,full not supported, and wildignorecase not worked for all cmd, and wildmenumode() seems might be wrong.

**Comment by wsdjeg** [comment]
Nice work, but How to deal with copyright issues? Neovim contributions since [b17d96](https://github.com/neovim/neovim/commit/b17d9691a24099c9210289f16afb1a498a89d803) are licensed under the Apache 2.0 license.

**Comment by vim-ml** [comment]
Hi,

On Tue, Feb 8, 2022 at 7:38 AM Wang Shidong ***@***.***>
wrote:

> Nice work, but How to deal with copyright issues? Neovim contributions
> since b17d96
> <https://github.com/neovim/neovim/commit/b17d9691a24099c9210289f16afb1a498a89d803>
> are licensed under the Apache 2.0 license.
>
>
>
The changes to support the command-line popup menu are made to the core Vim
command line completion code that Neovim uses from Vim. So I think that the
Vim
license covers these changes. But I am not a lawyer, so I don't know all
the nuances
with the license.

Regards,
Yegappan

**Comment by dpelle** [comment]
Using vim-8.2.4327, I get the pum completion on Ex line to work with:
```
vim --clean -c 'set wildoptions=pum'
```
It's very nice BTW. However, it does not work with:
```
vim --clean -c 'set wildoptions=pum wildmode=longest,list'
```
Any reason for pum completion to not work in that case?

**Comment by vim-ml** [comment]
Hi,

On Tue, Feb 8, 2022 at 8:02 AM Dominique Pellé ***@***.***>
wrote:

> Using vim-8.2.4327, I get the pum completion on Ex line to work with:
>
> vim --clean -c 'set wildoptions=pum'
>
> It's very nice BTW. However, it does not work with:
>
> vim --clean -c 'set wildoptions=pum wildmode=longest,list'
>
> Any reason for pum completion to not work in that case?
>
>
> Yes. Using the 'list' value in 'wildmode' should list all the matches but
not
use the wildmenu. As the help for 'wildmode' describes, if you use
"longest,full" or "longest:full", then it should open the wild menu.

Regards,
Yegappan

**Comment by Shane-XB-Qian** [comment]
> How to deal with copyright issues? Neovim contributions since b17d96<https://github.com/neovim/neovim/commit/b17d9691a24099c9210289f16afb1a498a89d803>are licensed under the Apache 2.0 license.


what's your point? is there any commercial code or brand issue or paper?


neovim looks porting vim patches everyday.

**Comment by bfredl** [comment]
@wsdjeg as the main author of the neovim implementation, I have previously stated that inclusion in vim under the vim license is okay.

**Comment by Shane-XB-Qian** [comment]
> Any reason for pum completion to not work in that case?


seems coz `wildmenu` didn't support others but `full`, so this/pum same.  :-)

// but i wish it can be working like `noinsert` option.

**Comment by dpelle** [comment]
Using this command, I see a regression:
```
vim --clean -c 'set wildoptions= wildmode=longest,list'
```
Then press for example `:c<tab>s`  and notice that the screen is cleared when pressing `s`.
I don't see this bug prior at 8.2.4324. The bug started to appear at 8.2.4325.

**Comment by dpelle** [comment]
> > However, it does not work with:
> > 
> >  vim --clean -c 'set wildoptions=pum wildmode=longest,list'
> >
> >Any reason for pum completion to not work in that case?
>
>  Yes. Using the 'list' value in 'wildmode' should list all the matches but
> not use the wildmenu. As the help for 'wildmode' describes, if you use
> "longest,full" or "longest:full", then it should open the wild menu.

I think it could be made to work with the pum though.  Maybe I'm missing
something, but when using the pum and I pressing `:ar`  I would like it to
complete to `:arg` (as all completions start with :arg) on the Ex line and
show the pum  with no item selected in the pum as it did not fully complete.

I also wonder, is there a way to always show the pum as I type an Ex
command without pressing <tab>?

**Comment by brammool** [comment]
> Using this command, I see a regression:
> ```
> vim --clean -c 'set wildoptions= wildmode=longest,list'
> ```
> Then press for example `:c<tab>s`  and notice that the screen is
> cleared when pressing `s`.  I don't see this bug prior at 8.2.4324.
> The bug started to appear at 8.2.4325.

I'll fix that.

-- 
"Making it up?  Why should I want to make anything up?  Life's bad enough
as it is without wanting to invent any more of it."
		-- Marvin, the Paranoid Android in Douglas Adams'
		   "The Hitchhiker's Guide to the Galaxy"

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by yegappan** [comment]
> > > However, it does not work with:
> > > vim --clean -c 'set wildoptions=pum wildmode=longest,list'
> > > Any reason for pum completion to not work in that case?
> > 
> > 
> > Yes. Using the 'list' value in 'wildmode' should list all the matches but
> > not use the wildmenu. As the help for 'wildmode' describes, if you use
> > "longest,full" or "longest:full", then it should open the wild menu.
> 
> I think it could be made to work with the pum though. Maybe I'm missing something, but when using the pum and I pressing `:ar<tab>` I would like it to complete to `:arg` (as all completions start with :arg) on the Ex line and show the pum with no item selected in the pum as it did not fully complete.
> 
>

We can support this but this will need some work. I will look into this later.

>
> I also wonder, is there a way to always show the pum as I type an Ex command without pressing ?
>

This will need quite a bit of work (to make it work like the insert-mode completion popup).

**Comment by wsdjeg** [comment]
> @wsdjeg as the main author of the neovim implementation, I have previously stated that inclusion in vim under the vim license is okay.

Ok, thank you. Reference at least should be mentioned in the Commit Message.

**Comment by Shane-XB-Qian** [comment]
@wsdjeg the reference is in this ticket description already, and implement logic looks based on vim core and vim wild menu.
it is just about upstream and fork.
// and BTW: what's related to your biz?

--
shane.xb.qian

**Pull request closed without merging**